pkg/addon-operator/ensure_crds.go

```diff
@@ -0,0 +1,225 @@
+
+	"github.com/flant/addon-operator/pkg/module_manager/models/modules"
+	"github.com/flant/addon-operator/sdk"
+	"github.com/flant/kube-client/client"
+)
+
+var crdGVR = schema.GroupVersionResource{
+	Group:    "apiextensions.k8s.io",
+	Version:  "v1",
+	Resource: "customresourcedefinitions",
+}
+
+func (op *AddonOperator) EnsureCRDs(module *modules.BasicModule) error {
+	crdsGlob := filepath.Join(module.GetPath(), "crds", "*.yaml")
+	result := new(multierror.Error)
+	cp, err := NewCRDsInstaller(op.KubeClient(), crdsGlob)
+	if err != nil {
+		result = multierror.Append(result, err)
+		return result
+	}
+
+	if cp == nil {
+		return nil
+	}
+
+	return cp.Run(context.TODO()).ErrorOrNil()
+}
+
+// CRDsInstaller simultaneously installs CRDs from specified directory
+type CRDsInstaller struct {
+	k8sClient    dynamic.Interface
+	crdFilesPath []string
+	buffer       []byte
+
+	// concurrent tasks to create resource in a k8s cluster
+	k8sTasks *multierror.Group
+}
+
+func (cp *CRDsInstaller) Run(ctx context.Context) *multierror.Error {
+	result := new(multierror.Error)
+
+	for _, crdFilePath := range cp.crdFilesPath {
+		if match := strings.HasPrefix(filepath.Base(crdFilePath), "doc-"); match {
+			continue
+		}
+
+		err := cp.processCRD(ctx, crdFilePath)
+		if err != nil {
+			err = fmt.Errorf("error occurred during processing %q file: %w", crdFilePath, err)
+			result = multierror.Append(result, err)
+			continue
+		}
+	}
+
+	errs := cp.k8sTasks.Wait()
+	if errs.ErrorOrNil() != nil {
+		result = multierror.Append(result, errs.Errors...)
+	}
+
+	return result
+}
+
+func (cp *CRDsInstaller) processCRD(ctx context.Context, crdFilePath string) error {
+	crdFileReader, err := os.Open(crdFilePath)
+	if err != nil {
+		return err
+	}
+	defer crdFileReader.Close()
+
+	crdReader := apimachineryYaml.NewDocumentDecoder(crdFileReader)
+
+	for {
+		n, err := crdReader.Read(cp.buffer)
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+
+			return err
+		}
+
+		data := cp.buffer[:n]
+		if len(data) == 0 {
+			// some empty yaml document, or empty string before separator
+			continue
+		}
+		rd := bytes.NewReader(data)
+		err = cp.putCRDToCluster(ctx, rd, n)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (cp *CRDsInstaller) putCRDToCluster(ctx context.Context, crdReader io.Reader, bufferSize int) error {
+	var crd *v1.CustomResourceDefinition
+
+	err := apimachineryYaml.NewYAMLOrJSONDecoder(crdReader, bufferSize).Decode(&crd)
+	if err != nil {
+		return err
+	}
+
+	// it could be a comment or some other peace of yaml file, skip it
+	if crd == nil {
+		return nil
+	}
+
+	if crd.APIVersion != v1.SchemeGroupVersion.String() && crd.Kind != "CustomResourceDefinition" {
+		return fmt.Errorf("invalid CRD document apiversion/kind: '%s/%s'", crd.APIVersion, crd.Kind)
+	}
+
+	if len(crd.ObjectMeta.Labels) == 0 {
+		crd.ObjectMeta.Labels = make(map[string]string, 1)
+	}
+	crd.ObjectMeta.Labels["heritage"] = "deckhouse"
```

Comment:
We should set it via env variables

Something like
`os.Getenv("ADDON_OPERATOR_CRD_EXTRALABELS")`
and set it in the deckhouse `ADDON_OPERATOR_CRD_EXTRALABELS=heritage=deckhouse,foo=bar`

Reply:
OK
but use `ADDON_OPERATOR_EXTRA_LABELS` variable